lib/nx_stencil.ex

```diff
@@ -1,51 +1,65 @@
 defmodule NxStencil do
-  def enumerate(lst) do
-    Enum.zip(lst, Stream.iterate(0, &(&1 + 1)))
+  defmodule StencilData do
+    @enforce_keys [
+      :up_rank,
+      :self_rank,
+      :down_rank,
+      :top_halo,
+      :self_data,
+      :bottom_halo
+    ]
+
+    defstruct @enforce_keys
   end
 
-  def access(nil, _, _), do: 0.0
-  def access(_, x, size) when x < 0 or x > size - 1, do: 0.0
-  def access(arr, x, _), do: Enum.at(arr, x)
+  defmodule InitData do
+    @enforce_keys [:parent, :data, :iters]
 
-  def len_or_nil(nil), do: "nil"
-  def len_or_nil(v), do: inspect(length(v))
+    defstruct @enforce_keys
+  end
+
+  defp access(nil, _, _), do: 0.0
+  defp access(_, x, size) when x < 0 or x > size - 1, do: 0.0
+  defp access(arr, x, _), do: Enum.at(arr, x)
```

Comment:
Iterating through the list, index access, is always going to be slow compared to pattern matching. Or even using an index map. If you're going to keep tweaking and tinkering with this, take a gander at [Benchee](https://hex.pm/packages/benchee)